processing/src/main/java/org/apache/druid/segment/join/lookup/LookupJoinMatcher.java

```diff
@@ -0,0 +1,313 @@
+  private final JoinConditionAnalysis condition;
+  private final List<Supplier<String>> keySuppliers;
+  private final ColumnSelectorFactory selectorFactory = new LookupColumnSelectorFactory(currentEntry::get);
+
+  // matchedKeys and matchingRemainder are used to implement matchRemainder().
+  private boolean matchingRemainder = false;
+  private final Set<String> matchedKeys;
+
+  private LookupJoinMatcher(
+      LookupExtractor extractor,
+      ColumnSelectorFactory leftSelectorFactory,
+      JoinConditionAnalysis condition,
+      @Nullable List<Expr> keyExprs,
+      boolean remainderNeeded
+  )
+  {
+    this.extractor = extractor;
+    this.matchedKeys = remainderNeeded && !condition.isAlwaysTrue() && !condition.isAlwaysFalse()
+                       ? new HashSet<>()
+                       : null;
+    this.condition = condition;
+
+    if (keyExprs != null) {
+      this.keySuppliers = keyExprs.stream()
+                                  .map(
+                                      expr ->
+                                          ColumnProcessors.makeProcessor(
+                                              expr,
+                                              ValueType.STRING,
+                                              LEFT_KEY_READER,
+                                              leftSelectorFactory
+                                          )
+                                  )
+                                  .collect(Collectors.toList());
+    } else {
+      // This check is to guard against bugs; users should never see it.
+      Preconditions.checkState(
+          condition.isAlwaysFalse() || condition.isAlwaysTrue(),
+          "Condition must be always true or always false when keySuppliers == null"
+      );
+
+      this.keySuppliers = null;
+    }
+
+    // Verify that extractor can be iterated when needed.
+    if (condition.isAlwaysTrue() || remainderNeeded) {
+      Preconditions.checkState(
+          extractor.canIterate(),
+          "Cannot iterate lookup, but iteration is required for this join"
+      );
+    }
+  }
+
+  public static LookupJoinMatcher create(
+      LookupExtractor extractor,
+      ColumnSelectorFactory leftSelectorFactory,
+      JoinConditionAnalysis condition,
+      boolean remainderNeeded
+  )
+  {
+    final List<Expr> keyExprs;
+
+    if (condition.isAlwaysTrue()) {
+      keyExprs = null;
+    } else if (condition.isAlwaysFalse()) {
+      keyExprs = null;
+    } else if (condition.getNonEquiConditions().isEmpty()
+               && condition.getEquiConditions()
+                           .stream()
+                           .allMatch(eq -> eq.getRightColumn().equals(LookupColumnSelectorFactory.KEY_COLUMN))) {
+      keyExprs = condition.getEquiConditions().stream()
+                          .map(Equality::getLeftExpr)
+                          .collect(Collectors.toList());
+    } else {
+      throw new IAE("Cannot join lookup with condition: %s", condition);
```

Comment:
Suggest splitting up the non-equiconditions check and the lookup key column check, and using a specific exception message for each case

Reply:
Ah, good idea!!